File: 0110-balanced-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def height(root: TreeNode) -> int:
            if root is None:
                return 0
            left_height = height(root.left)
            if left_height == -1:
                return -1
            right_height = height(root.right)
            if right_height == -1:
                return -1
            if abs(left_height - right_height) > 1:
                return -1
            return 1 + max(left_height, right_height)
        return height(root) != -1

File: 0110-balanced-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    var height func(*TreeNode) int
    height = func (root *TreeNode) int {
        if root == nil {
            return 0
        }
        l_height := height(root.Left)
        if l_height < 0 {
            return -1
        }
        r_height := height(root.Right)
        if r_height < 0 {
            return -1
        }
        if 2 * max(l_height, r_height) - (l_height + r_height) > 1 {
            return -1
        }

        return 1 + max(l_height, r_height)
    }
    return height(root) != -1
}

func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

